src/member-api.js

```diff
@@ -54,42 +75,148 @@ export class MemberApi extends TypedEmitter {
+      const { name: invitorName } = await this.getById(this.#ownDeviceId)
+      // since we are always getting #ownDeviceId,
+      // this should never throw (see comment on getById), but it pleases ts
+      if (!invitorName)
+        throw new Error(
+          'Internal error trying to read own device name for this invite'
+        )
+      handleAborted()
 
-    const response = await this.#rpc.invite(deviceId, {
-      projectKey: this.#projectKey,
-      encryptionKeys: this.#encryptionKeys,
-      projectInfo: { name: project.name },
-      roleName: roleName || ROLES[roleId].name,
-      roleDescription,
-      invitorName: deviceName,
-      timeout,
-    })
+      const projectId = projectKeyToId(this.#projectKey)
+      const projectPublicId = projectKeyToPublicId(this.#projectKey)
+      const project = await this.#dataTypes.project.getByDocId(projectId)
+      const projectName = project.name
+      if (!projectName)
+        throw new Error('Project must have a name to invite people')
+      handleAborted()
 
-    if (response === InviteResponse_Decision.ACCEPT) {
-      await this.#roles.assignRole(deviceId, roleId)
-    }
+      const inviteId = crypto.randomBytes(32)
+
+      const {
+        promise: receiveInviteResponsePromise,
+        resolve: gotInviteResponse,
+      } = /** @type {typeof promiseWithResolvers<InviteResponse>} */ (
+        promiseWithResolvers
+      )()
+      /**
+       * @param {string} peerId
+       * @param {InviteResponse} inviteResponse
+       */
+      const onInviteResponse = (peerId, inviteResponse) => {
+        const isResponseToThisInvite =
+          timingSafeEqual(peerId, deviceId) &&
+          timingSafeEqual(inviteId, inviteResponse.inviteId)
+        if (!isResponseToThisInvite) return
+
+        gotInviteResponse(inviteResponse)
+        this.#rpc.off('invite-response', onInviteResponse)
+      }
+      this.#rpc.on('invite-response', onInviteResponse)
+      cleanupFns.push(() => {
+        this.#rpc.off('invite-response', onInviteResponse)
+      })
+
+      await this.#rpc.sendInvite(deviceId, {
+        inviteId,
+        projectPublicId,
+        projectName,
+        roleName,
+        roleDescription,
+        invitorName,
+      })
+
+      handleAborted()
 
-    return response
+      const inviteResponse = await receiveInviteResponsePromise
+
+      handleAborted()
+
+      switch (inviteResponse.decision) {
+        case InviteResponse_Decision.ALREADY:
+        case InviteResponse_Decision.REJECT:
+          return inviteResponse.decision
+        case InviteResponse_Decision.UNRECOGNIZED:
+          return InviteResponse_Decision.REJECT
+        case InviteResponse_Decision.ACCEPT:
+          handleAborted()
+
+          // We should assign the role locally *before* sharing the project details
+          // so that they're part of the project even if they don't receive the
+          // project details message.
```

Comment:
Calling this out, as I'm not sure it's correct.

Reply:
I think this is ok, although also not entirely sure. If we wait until after we don't know whether they have maybe received the join details, but we lost the connection before we learned that, which would leave them "in the project" but without a role, which would be a bad experience, so I think the order here is the safest thing to do.